cpp/src/arrow/util/bit-util.h

```diff
@@ -296,6 +297,25 @@ static inline int Log2(uint64_t x) {
   return result;
 }
 
+/// \brief Count the number of leading zeros in a 32 bit integer.
+static inline int64_t CountLeadingZeros(uint32_t value) {
+  DCHECK_NE(value, 0);
+#if defined(__clang__) || defined(__GNUC__)
+  return static_cast<int64_t>(__builtin_clz(value));
```

Comment:
Interesting tidbit here is that `__builtin_clz` compiles to `31 - bit scan reverse` on gcc and clang.